mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -830,17 +830,617 @@ LogicalResult ThreadwiseWriteAllRewritePattern::matchAndRewrite(
+// BlockwiseReduce lowering.
+//===----------------------------------------------------------------------===//
+
+struct BlockwiseReduceRewritePattern
+    : public OpConversionPattern<BlockwiseReduceOp> {
+  using OpConversionPattern<BlockwiseReduceOp>::OpConversionPattern;
+
+  int64_t calculateNonReductionDimProduct(ArrayRef<int64_t> toReduceShape,
+                                          int64_t axis) const {
+    int64_t dimProduct = 1;
+    for (size_t i = 0; i < toReduceShape.size(); i++) {
+      if (i != (size_t)axis) {
+        dimProduct *= toReduceShape[i];
+      }
+    }
+    return dimProduct;
+  }
+
+  // This function will append views to target a flat LDS buffer
+  // where non-reduction dims are laid contigously as they are expected
+  // function on parallel.
+  ArrayAttr createLDSWorkspaceView(Location loc, PatternRewriter &rewriter,
+                                   ArrayAttr regTensorView, int64_t reduceAxis,
+                                   bool makeRDimZero = false) const {
+
+    TransformMapAttr lowestTr =
+        regTensorView[regTensorView.size() - 1].cast<TransformMapAttr>();
+    ArrayRef<int64_t> lowestShape = lowestTr.getLowerBounds().asArrayRef();
+    TopDownTMBuilder tensorToLDSViewBuilder(rewriter, lowestShape, loc);
+    SmallVector<StringRef, 4> upperNameRefs;
+    tensorToLDSViewBuilder.getStartNames(upperNameRefs);
+
+    int64_t nonReduceMergeDimSize = 1;
+    SmallVector<StringRef, 4> nonReduceNameRefs;
+    SmallVector<unsigned, 4> nonReduceDims;
+    SmallVector<int64_t, 4> nonReduceDimSizes;
+    for (auto [dim, dimSize] : llvm::enumerate(lowestShape)) {
+      if (dim != (size_t)reduceAxis) {
+        nonReduceMergeDimSize *= dimSize;
+        nonReduceNameRefs.push_back(upperNameRefs[dim]);
+        nonReduceDims.push_back(dim);
+        nonReduceDimSizes.push_back(dimSize);
+      }
+    }
+    tensorToLDSViewBuilder.unmerge("nrDim", 1, nonReduceNameRefs,
+                                   nonReduceDimSizes);
+    if (makeRDimZero) {
+      tensorToLDSViewBuilder.constDim("rDim", 0, 0, lowestShape[reduceAxis]);
+    } else {
+      tensorToLDSViewBuilder.passThrough({"rDim"}, {0},
+                                         {upperNameRefs[reduceAxis]});
+    }
+    TransformMapAttr twoDimLDSView = tensorToLDSViewBuilder.get();
+
+    TopDownTMBuilder flatLDSViewBuilder =
+        TopDownTMBuilder::below(tensorToLDSViewBuilder, twoDimLDSView);
+    flatLDSViewBuilder.unmerge(
+        "flatDim", 0, {"nrDim", "rDim"},
+        {nonReduceMergeDimSize, lowestShape[reduceAxis]});
+    TransformMapAttr flatLDSView = flatLDSViewBuilder.get();
+    SmallVector<Attribute> threadsToLDSViewAttrs;
+    for (Attribute trMap : regTensorView) {
+      threadsToLDSViewAttrs.push_back(trMap);
+    }
+    threadsToLDSViewAttrs.push_back(twoDimLDSView);
+    threadsToLDSViewAttrs.push_back(flatLDSView);
+    return rewriter.getArrayAttr(threadsToLDSViewAttrs);
```

Comment:
Instead of doing this, could you add some `rock.transform` operations and then `untransform()` later? Or would that make things uglier?

Reply:
I dont have a preference either way in terms of clarity -- I chose this because I felt it was unnecessary to re-traverse the IR (i.e. do untransform) within the same lowering pass. I really dont have an opinion between passing the ArrayAttr of transforms vs a stack of transform ops apart from the latter has to spend marginally more cycles compiling.